routers/api/packages/api.go

```diff
@@ -167,6 +168,49 @@ func CommonRoutes(ctx gocontext.Context) *web.Route {
 				})
 			})
 		}, reqPackageAccess(perm.AccessModeRead))
+		r.Group("/conda", func() {
+			var (
+				downloadPattern = regexp.MustCompile(`\A(.+/)?(.+)/((?:[^/]+(?:\.tar\.bz2|\.conda))|(?:current_)?repodata\.json(?:\.bz2)?)\z`)
+				uploadPattern   = regexp.MustCompile(`\A(.+/)?([^/]+(?:\.tar\.bz2|\.conda))\z`)
+			)
+
+			r.Route("/*", "GET,PUT", func(ctx *context.Context) {
+				path := ctx.Params("*")
+
+				if ctx.Req.Method == "GET" {
+					m := downloadPattern.FindStringSubmatch(path)
+					if len(m) > 0 {
+						ctx.SetParams("channel", strings.TrimSuffix(m[1], "/"))
+						ctx.SetParams("architecture", m[2])
+						ctx.SetParams("filename", m[3])
+
+						switch m[3] {
+						case "repodata.json", "repodata.json.bz2", "current_repodata.json", "current_repodata.json.bz2":
+							conda.EnumeratePackages(ctx)
+						default:
+							conda.DownloadPackageFile(ctx)
+						}
+						return
+					}
+				} else {
+					reqPackageAccess(perm.AccessModeWrite)(ctx)
+					if ctx.Written() {
+						return
+					}
+
+					m := uploadPattern.FindStringSubmatch(path)
+					if len(m) > 0 {
+						ctx.SetParams("channel", strings.TrimSuffix(m[1], "/"))
+						ctx.SetParams("filename", m[2])
+
+						conda.UploadPackageFile(ctx)
+						return
+					}
+				}
+
+				ctx.Status(http.StatusNotFound)
+			})
+		}, reqPackageAccess(perm.AccessModeRead))
```

Comment:
The flow of control here is somewhat convoluted.

I have a feeling you should have `r.Get` and a `r.Put` methods and you should convert your `if` statements to return quickly:

```go
r.Get("/*",  func(ctx *context.Context) {
	path := ctx.Params("*")
	m := downloadPattern.FindStringSubmatch(path)
	if len(m) == 0 {
		ctx.Status(http.StatusNotFound)
		return
	}

	.... handle the download ...
```

Reply:
Yes, that's better 👍